File: level1/98/binarySearch.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def dfs(self, root, min, max):
        if root == None:
            return True
        if root.val > min and root.val < max:
            return self.dfs(root, min, root.val) and self.dfs(root, root.val, max)
        return False
    def isValidBST(self, root):
        if root == None:
            return True
        if root and root.left == None and root.right == None:
            return True
        return self.dfs(root,-1e12, 1e12)
        
            

File: level1/98/binarySearch.c
 #include <stdio.h>
 #include <stdbool.h>

 struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
 };

bool depthFirstSearch(struct TreeNode* root, long long int min, long long int max) {
    if (root == NULL)
        return true;
    if (root->val > min && root->val < max)
        return depthFirstSearch(root->left, min, root->val) && depthFirstSearch(root->right, root->val, max);
    return false;
}

bool isValidBST(struct TreeNode* root) {
    if (root == NULL)
        return true;
    if (root && !root->left && !root->right)
        return true;
    return depthFirstSearch(root, -1e12, 1e12);
}